### ISSUE TITLE
Trying to analyse game from WCC today

### ISSUE THREAD

**Issue by jjoshua2** [BrentBaccala/hoffman#1]
Trying to analyse fen using 7 man syzygy.
`5k2/8/5pK1/5PbP/2Bn4/8/8/8 b - - 8 68`
I tried `perl genctlfile.pl kbnpvkbpp.xml`
And took the kbnpvkbpp.xml and renamed all the endings to syzygy.
```
./hoffman -g .\wcc.xml -t 4 -P 5000
Hoffman Version 1030
Total indices: 72742152954240
BUG: Identical pieces don't have the same number of total semilegal positions
BUG: Identical pieces don't have the same number of encoding values
```

**Comment by jjoshua2** [comment]
I also tried and get similar error:
```
<tablebase>
   <prune-enable color="white" type="concede"/>
   <prune-enable color="black" type="concede"/>
   <dtc/>
   <piece color="white" type="king"/>
   <piece color="black" type="king"/>
   <piece color="white" type="pawn" location="f5+"/>
   <piece color="white" type="pawn" location="h5+"/>
   <piece color="white" type="bishop" />
   <piece color="black" type="pawn" location="f6+"/>
   <piece color="black" type="knight" />
   <piece color="black" type="bishop" />
   <prune color="white" type="concede" move="P=?"/>
   <prune color="black" type="concede" move="P=?"/>
   <output filename="wcc2018.htb"/>
</tablebase>
```

**Comment by BrentBaccala** [comment]
How to analyze game 6 from the 2018 world championship?

Good question!

You're on the right track with your posted XML configuration; here are the issues that I see:

1. Syzygy tablebases don't store results for every position; sometimes you have to build a search tree, look at all possible positions that can arise from your problem position, and then probe the Syzygy tablebase for all of them in order to figure out the result.  That's all built into the Syzygy code.  The problem is that you make have to search past captures and pawn promotions, so to use the "KBPPKNP" tablebase, for example, you need the "KBPKNP", "KNPKPP", "KBPPKN", and "KBPPKP"  to handle the captures, plus all the capture possibilities that arise from them, plus all the pawn promotion possibilities.  In short, Syzygy tablebases are somewhat impractical to use with Hoffman because you need so many of them.  If you've got about 20 TB of disk space, you're in luck.  Otherwise, you have to use Nalimov tablebases, and it's a considerably more complicated calculation.

2. Assuming that you've got a full set of Syzygy TBs available (a big if), then your XML code is close.  Use pawngen instead of plus-pawns (pawngen is newer and better).  Use "P*=*" for your prune statements to match capture-promotions as well as non-capture promotions.  Consider replacing `<dtc/> `with`<basic/>` to make your output file smaller.

I took a quick look, and it only requires about 8.6 GB of RAM for the calculation.  A reasonably souped up home PC could do it... if you've got a full set of Syzygy's handy.

Here's the control file that I would use:

```
<tablebase>
   <prune-enable color="white" type="concede"/>
   <prune-enable color="black" type="concede"/>
   <dtc/>
   <piece color="white" type="king"/>
   <piece color="black" type="king"/>
   <piece color="white" type="bishop" />
   <piece color="black" type="knight" />
   <piece color="black" type="bishop" />

   <pawngen white-pawn-locations="f5 h5" black-pawn-locations="f6"
	    white-pawns-required="2" black-pawns-required="1"/>

   <futurebase filename="KBNPvKBP.rtbw" colors="invert"/>
   <futurebase filename="KBNPvKPP.rtbw" colors="invert"/>
   <futurebase filename="KBPPvKNP.rtbw"/>
   <futurebase filename="KBPPvKBP.rtbw"/>
   <futurebase filename="KBPPvKBN.rtbw"/>

   <prune color="white" type="concede" move="P*=*"/>
   <prune color="black" type="concede" move="P*=*"/>

   <output filename="wcc2018.htb"/>

</tablebase>
```

**Comment by jjoshua2** [comment]
Got 32gb of memory and all 7 man wdl but not dtz. I'll try this later
today. I'm curious if this would also work for when it initially hit 8 men?
If i could just do that and use it for all positions after that it could be
more efficient? Not sure if I can easily go backwards and use existing work
otherwise.

**Comment by jjoshua2** [comment]
I tried using your file exactly, just edited the filenames to be the location like ```<futurebase filename="C:\Chess\tb\KBPPvKBP.rtbw"/>
   <futurebase filename="E:\KBPPvKBN.rtbw"/>```
BTW it would be VERY nice if you could have a list of paths where it would search syzygy. It seems you have it only for nalimov according to help.
```./hoffman -g .\wcc2018.xml -t 16
Hoffman Version 1030
Total indices: 4537331280
BUG: Identical pieces don't have the same number of total semilegal positions
BUG: Identical pieces don't have the same number of encoding values
```

I took out the full paths and got a different error...
 ```./hoffman -g .\wcc2018.xml -t 16
Hoffman Version 1030
Total indices: 4537331280
34 possible WHITE futuremoves
10 possible BLACK futuremoves
Too many futuremoves - 34!  (only 32 bits futurevector_t)
```
Also tried `-n "C:\chess\tb;E:\F:\D:\tb\tb;G:\"` and still didn't work. (In case Nalimov path works for syzygy too)
Also tried `perl.exe .\pawngen .\wcc2018.xml`

```
Can't locate XML/LibXML.pm in @INC (you may need to install the XML::LibXML module) (@INC contains: /usr/lib/perl5/site_perl /usr/share/perl5/site_perl /usr/lib/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib/perl5/core_perl /usr/share/perl5/core_perl .) at .\pawngen line 17.
BEGIN failed--compilation aborted at .\pawngen line 17.
```

**Comment by BrentBaccala** [comment]
That's a known bug in Hoffman 1.7.  It affects Nalimov and Syzygy tablebases with pawns.  Can you use the current development version from github?

I'm planning a 1.8 release, but I've got some new features that I want to test out before I make it official.

**Comment by jjoshua2** [comment]
Well I tried using Ubuntu 18 for Windows subsystem and couldn't get it to
make or autoconf. What command do u run exactly? Or I'll test a dev exe

**Comment by jjoshua2** [comment]
```josh@DESKTOP-34OEHFL:~/hoffman$ make
make: *** No targets specified and no makefile found.  Stop.
josh@DESKTOP-34OEHFL:~/hoffman$ autoreconf -fi
configure.ac:24: error: possibly undefined macro: AC_DEFINE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:38: error: possibly undefined macro: AS_IF
configure.ac:44: error: possibly undefined macro: AC_MSG_WARN
autoreconf: /usr/bin/autoconf failed with exit status: 1
```
Or using the one from the readme:
```
josh@DESKTOP-34OEHFL:~/hoffman$ CFLAGS="-march=core2 -O3" ./configure
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
./configure: line 2370: AX_CXX_COMPILE_STDCXX_11: command not found
./configure: line 2371: AX_CXXFLAGS_WARN_ALL: command not found
./configure: line 2386: syntax error near unexpected token `CHECK_INTEL_ATOMIC_OPS'
./configure: line 2386: `  AX_PTHREAD(CHECK_INTEL_ATOMIC_OPS'
```

**Comment by jjoshua2** [comment]
Got as far as this:
```
josh@DESKTOP-34OEHFL:~/hoffman$ autoreconf -f
josh@DESKTOP-34OEHFL:~/hoffman$ CFLAGS="-march=native -O3" ./configure
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking whether g++ supports C++11 features by default... yes
checking CXXFLAGS for maximum warnings... -Wall
configure: error: cannot find install-sh, install.sh, or shtool in "." "./.." "./../.."
```

**Comment by jjoshua2** [comment]
I got someone to help me by copying the default install-sh and cloning fathom. You really should include these things in your readme :)
I think its creating now. I get the same 8.6gb requirement you mentioned. It looks like its pruning a bunch of captures, but given I already have all the 7 man syzygy I don't think I should prune any of them?
Thanks!

**Comment by BrentBaccala** [comment]
I'm glad you got it working.  I'm still trying to update my Windows 10 virtual machine so that I can attempt to compile it using Ubuntu for Windows.

The problems you had make sense to me.  I'll try to improve the documentation.

The messages are a bit deceiving.  It's not really "pruning" those captures; it's just identified that they're all being handled by the futurebases, so they don't need to be tracked.

Did you give a `-t` option?  It should speed things up a bit if you tell it to use the same number of threads as processor cores you've got available.

Let me know how it goes.  I'd like to get a copy of your results, since I don't have the TBs of disk space needed for Syzygy's to do the calculation myself.

**Comment by jjoshua2** [comment]
Yes, I'm doing -t 4. Any idea how long it should take? Hours, days weeks? I've got this so far.
```
0 unpruned WHITE futuremoves
0 unpruned BLACK futuremoves
46 maximum white moves; 42 maximum black moves
Malloced 8654MB for tablebase entries
Entries format: 6 bits movecnt; 10 bits dtm
Initializing tablebase
...................................
```
Also what does 0 unpruned futuremoves mean?

And could I use the database to calculate from before 64. H5 by using these and just regenerating the small amount extra of starting pawn on H4?

**Comment by BrentBaccala** [comment]
Maybe ten hours or so?  I doubt it will take more than a day.  I'm not sure if the Syzygy tablebases will slow things down, because I've hardly used them at all.

Actually, now that I think about it, it might take longer.  More like a day.  You're backproping from five 7 piece tablebases.  That might take a while.

I'm guessing, really.

The "0 unpruned futuremoves" statement just means that it knows that all possibles moves out of this tablebase have been handled by either futurebases or pruning statements, so it doesn't have to track them individually.

As far as calculating with the pawn on h4, that's not currently supported.  In principle it's possible, and might not be too difficult to implement.  I'll think about it some more.

**Comment by jjoshua2** [comment]
I don't need them until the probing right? I didn't put in the full path in
the xml yet

**Comment by BrentBaccala** [comment]
No, you need them for the creation step, and it doesn't check at first to see if it can access all of the files that it needs.  You might get an error when it starts trying to back propagate from those Syzygy files.

If you don't want to restart from scratch, I'd suggest putting symbolic links to all of your Syzygy files (use the `mklink` command on Windows) in the directory that hoffman is running in.  I do mean **all** of your Syzygy files, not just the ones referenced in the XML.

**Comment by jjoshua2** [comment]
Yah it crashes. What do i do on Linux? I tried a ln -s but it made a folder
named f of my f drive instead of putting all the files in . Where i think
it need to be.

**Comment by BrentBaccala** [comment]
Maybe try it with full pathnames to the Syzygy files, if they're all in one directory.

**Comment by jjoshua2** [comment]
They are in 5 different folders on 5 different drives. I think maybe this will work
```
ln -s /mnt/f/* . 
```

**Comment by BrentBaccala** [comment]
OK, I see now.  Looks good.  (I've never used Ubuntu on Windows before)

If the files are in the root directory of F drive, then you've got the right symlink command.  Do that five times for each of your five drives, and you should have symlinks to all of your Syzygy files, and hoffman should work fine.

Maybe I should add an option to set the search path for Syzygy files.

**Comment by jjoshua2** [comment]
I think that's needed since `ls` can't seem to handle this many files it seems very well. And having this many different symlinks is very confusing.

**Comment by jjoshua2** [comment]
```
Back propagating from 'KBNPvKBP.rtbw'
TBMAX_PIECE limit too low!
```
I see `#define TBPIECES 7` though in Fathom/src/tbcore.h

You probably need to adjust some of these. 
tbcore.c:#define TBMAX_PIECE 254
tbcore.c:#define TBMAX_PAWN 256

And some 7 pieces use 16 bit something different. So Fathom probably needs some more extensive changes...

**Comment by BrentBaccala** [comment]
I added a new `-S` option to set the search path for Syzygy tablebases (you can get it with a `git pull`), but it sounds like we've got some other problems with Fathom.

Try increasing `TBMAX_PIECE` and `TBMAX_PAWN`.  It looks like those are the sizes of tables of the files.  You've probably got more than 254 pawn-less Syzygy files, and i've never tested it with that many.

I have tested it with a half dozen files or so, so I believe that the 7-piece probe code is basically correct.

You can use Hoffman to directly probe Syzygy tablebases, if you want to test it that way.

**Comment by BrentBaccala** [comment]
I've also added a new `-U` option to set the size of the "unpropagated index table".  This feature isn't used for DTC tablebases, but it can really speed up a bitbase (i.e, win-draw-loss) calculation.  If you want to try it, `git pull` to get the latest version of the code, change `<dtc/>` to `<basic/>` in the XML and set `-U` in megabytes to be as large as possible.

We know that the game 6 calculation will use 8.6 GB for the tablebase, so if you've got 16 GB of RAM total, using `-U 1024` will make that a 1 GB table.  If you're not using the computer for much else, you might even bump it up to 4 GB with `-U 4096`

**Comment by jjoshua2** [comment]
Can you just take this file? It's by syzygy author...
https://github.com/syzygy1/Cfish/blob/7men/src/tbcore.c
```
#define TBMAX_PIECE 650
#define TBMAX_PAWN 861
#define HSHMAX 10
```
Also grab the tbcore.h next to it...

**Comment by BrentBaccala** [comment]
The advantage of using Jon Dart's version of this code is that it has no Stockfish dependencies, and therefore no GPL licensing issues.  I tried dropping in Roland de Man's version, like you suggested, but it didn't compile.  I'd suggest staying with Dart's code and adjusting those three `#define` constants to be 650, 861, and 10, like in Roland's version.  It works for me, but I don't have a full set of 7's, so I can't be sure if it will work for you.

**Comment by jjoshua2** [comment]
Ronald's version is from scratch not from Stockfish, and has MIT license. He's the author of syzygy, it doesn't come from SF. `This file may be redistributed and/or modified without restrictions.`

**Comment by jjoshua2** [comment]
I pulled both the tbcore.c and tbcore.h and it seems to compile for me... Did you just do one? It still crashes with same error about tbmax_piece...

**Comment by jjoshua2** [comment]
7 man dtz ` DTZ format has to be adapted to allow 16-bit values in the "map"' array` http://talkchess.com/forum3/viewtopic.php?f=7&t=66797&start=20#p754231
You should test your probing code on one of the files that needs this to work. I'm quite sure you will find it either doesn't work or has corrupted values.

Ahah! here syzygy says what you need! http://talkchess.com/forum3/viewtopic.php?f=7&t=66797&p=763890&hilit=fathom#p763890

**Comment by BrentBaccala** [comment]
I contacted Jon Dart about updating Fathom to handle 7 piece tablebases, and here is the relevant portion of his reply:

> It hasn't been integrated into Fathom yet. When I compared the CFish  code to the Fathom code, I got 3000 lines of diffs. So that is not a  small job to integrate. I'd like to do it, but it's not something I have time for right now, and then of course it would need testing.

So... it looks like 7 piece support will be slower in coming that I had thought.  I do not have enough disk space to do meaningful testing, so it might have to wait for somebody else to update Fathom.

Another possibility would be to analyse the game using six piece Nalimov tablebases.
